src/dune_rules/ctypes_rules.ml

```diff
@@ -0,0 +1,117 @@
+   - generate a functions generator
+   - set up a discovery program to query pkg-config for compile and link flags
+   - use the types/data and functions modules you filled in to tie everything
+   together into a neat library
+*)
+
+let gen_rule ~targets ~action () =
+  let open Dune_file in
+  { Rule.targets; action }
+
+let gen_library ?wrapped ?foreign_stubs ?c_library_flags ~name
+      ~public_name ~modules ~libraries () =
+  let open Dune_file in
+  { Library.name; public_name; modules; libraries; wrapped; foreign_stubs }
+
+let gen_executable ~name ~modules ~libraries () =
+  let open Dune_file in
+  { Executable.name; modules; libraries }
+
+(* It may help to understand what this generator function is trying to do by
+   having a look at the hand-written version it's replacing.
+   XXX: link to mpg123 dune file *)
+let really_expand lib ctypes =
+  let open Dune_file in
+  [ gen_executable
+      ~name:"mpg123_discover"
+      ~libraries:["dune.configurator"]
+      ()
+  ; gen_rule
+      ~targets:["c_flags.sexp"; "c_flags.txt"; "c_library_flags.sexp"]
+      ~action:["run discover.exe"]
+      ()
+  ; gen_library
+      ~name:("mpg123_c_type_descriptions")
+      ~public_name:("mpg123.c_type_descriptions")
+      ~modules:"Mpg123_c_type_descriptions"
+      ~libraries:["ctypes"]
+      ()
+  ; gen_executable
+      ~name:"type_gen"
+      ~modules:"Type_gen"
+      ~libraries:["ctypes.stubs"; "ctypes.foreign"; "mpg123_c_type_descriptions"]
+      ()
+  ; gen_rule_stdout
+      ~with_stdout_to:"c_generated_types.c"
+      ~run:("./type_gen.exe")
+      ()
+  ; gen_rule
+      ~targets:["c_generated_types.exe"]
+      ~deps:[":c c_generated_types.c"]
+      ~action:["system blah blah"]
+      ()
+  ; gen_rule_stdout
+      ~with_stdout_to:"mpg123_c_generated_types.ml"
+      ~run:("./c_generated_types.exe")
+      ()
+  ; gen_library
+      ~name:"mpg123_c_function_descriptions"
+      ~public_name:"mpg123.c_function_descriptions"
+      ~modules:["Mpg123_c_generated_types"; "Mpg123_c_function_descriptions";
+                "Mpg123_c_types"]
+      ~wrapped:false
+      ~flags:":standard -w -27 -w -9"
+      ~libraries:["ctypes"; "mpg123_c_type_descriptions"]
+      ()
+  ; gen_executable
+      ~name:"function_gen"
+      ~modules:"Function_gen"
+      ~libraries:["ctypes.stubs"; "mpg123_c_function_descriptions"]
+      ()
+  ; gen_rule_stdout
+      ~with_stdout_to:"c_generated_functions.c"
+      ~run:"./function_gen.exe c mpg123_stub"
+      ()
+  ; gen_rule_stdout
+      ~with_stdout_to:"mpg123_c_generated_functions.ml"
+      ~run:"./function_gen.exe ml mpg123_stub"
+      ()
+  ; gen_library
+      ~name:"mpg123_c"
+      ~public_name:"mpg123.c"
+      ~libraries:["ctypes"; "mpg123_c_function_descriptions"]
+      ~modules:["Mpg123_c"; "Mpg123_c_generated_functions"]
+      ~foreign_stubs:[
+        ("language" , "c");
+        ("names"    , "c_generated_functions");
+        ("flags"    , ":include c_flags.sexp")
```

Comment:
Why does this need to be in a separate library? Can't it be a part of the user defined library? I think I forgot why we need this `mpg123` vs. `mpg123.c` split.

Reply:
This is actually what luv does, which I adapted.  I didn't really question it but yeah, I'm not seeing why it can't simply set up in the library that contains the `ctypes` stanza.